src/interpreter.rs

```diff
@@ -90,51 +107,95 @@ fn eval_expr(expr: &Expr, env: &Env) -> Result<Value, String> {
+        Ok((Value::Num(ln), Value::Num(rn))) => Ok((Value::Num(ln / rn), env.clone())),
         Ok(_) => Err(String::from("Can only divide numers.")),
         Err(err) => Err(err),
     }
 }
 
-fn eval_expr_mul(lexpr: &Expr, rexpr: &Expr, env: &Env) -> Result<Value, String> {
+fn eval_expr_mul(lexpr: &Expr, rexpr: &Expr, env: &Env) -> Result<(Value, Env), String> {
     match eval_expr2(lexpr, rexpr, env) {
-        Ok((Value::Num(ln), Value::Num(rn))) => Ok(Value::Num(ln * rn)),
+        Ok((Value::Num(ln), Value::Num(rn))) => Ok((Value::Num(ln * rn), env.clone())),
         Ok(_) => Err(String::from("Can only multiply numers.")),
         Err(err) => Err(err),
     }
 }
 
-fn eval_expr_add(lexpr: &Expr, rexpr: &Expr, env: &Env) -> Result<Value, String> {
+fn eval_expr_add(lexpr: &Expr, rexpr: &Expr, env: &Env) -> Result<(Value, Env), String> {
     match eval_expr2(lexpr, rexpr, env) {
-        Ok((Value::Num(ln), Value::Num(rn))) => Ok(Value::Num(ln + rn)),
-        Ok((Value::Str(ls), Value::Str(rs))) => Ok(Value::Str(format!("{}{}", ls, rs))),
+        Ok((Value::Num(ln), Value::Num(rn))) => Ok((Value::Num(ln + rn), env.clone())),
+        Ok((Value::Str(ls), Value::Str(rs))) => {
+            Ok((Value::Str(format!("{}{}", ls, rs)), env.clone()))
+        }
         Ok(_) => Err(String::from("Can only add values of the same type.")),
         Err(err) => Err(err),
     }
 }
 
-fn eval_expr_sub(lexpr: &Expr, rexpr: &Expr, env: &Env) -> Result<Value, String> {
+fn eval_expr_sub(lexpr: &Expr, rexpr: &Expr, env: &Env) -> Result<(Value, Env), String> {
     match eval_expr2(lexpr, rexpr, env) {
-        Ok((Value::Num(ln), Value::Num(rn))) => Ok(Value::Num(ln - rn)),
+        Ok((Value::Num(ln), Value::Num(rn))) => Ok((Value::Num(ln - rn), env.clone())),
         Ok(_) => Err(String::from("Can only subtract numbers.")),
         Err(err) => Err(err),
     }
 }
 
-fn eval_expr_mod(lexpr: &Expr, rexpr: &Expr, env: &Env) -> Result<Value, String> {
+fn eval_expr_mod(lexpr: &Expr, rexpr: &Expr, env: &Env) -> Result<(Value, Env), String> {
     match eval_expr2(lexpr, rexpr, env) {
-        Ok((Value::Num(ln), Value::Num(rn))) => Ok(Value::Num(ln % rn)),
+        Ok((Value::Num(ln), Value::Num(rn))) => Ok((Value::Num(ln % rn), env.clone())),
         Ok(_) => Err(String::from("Can only mod numbers.")),
         Err(err) => Err(err),
     }
 }
+
+fn eval_command(expr: &Expr, env: &Env) -> Result<(Value, Env), String> {
+    if let Expr::Cmd(box Expr::Path(s), args) = expr {
+        let mut command = Command::new(s);
+        match args
+            .iter()
+            .map(|a| {
+                match eval_expr(a, env) {
+                    Ok((val, _env)) => Ok(val), //throwing away any env changes that occur here.
+                    Err(err) => Err(err),
+                }
+            })
+            .collect::<Result<Vec<Value>, String>>()
+        {
+            Ok(vals) => {
+                match command
```

Comment:
Doesn't really make sense to use `mut command` here.  We should avoid `mut` as much as we can, so I would replace `command` with its value (`Command::new(s)`).

Reply:
Heard